### PR TITLE
OMEMO - trust mode

### PR DESCRIPTION
**Pull request by StefanKropp**
**OMEMO Trust mode feature**

Added a new subcommand `trustmode` for OMEMO.

`/omemo trustmode manual|firstusage|blind`

The user can request the current mode via

`/omemo trustmode`

Example: Current trust mode is firstusage

The mode is stored as profanity preference.

```
[omemo]
trustmode=firstusage
```

1. trustmode manual  - Set the global OMEMO trust mode to manual, OMEMO keys has to be trusted manually.
2. trustmode firstusage - Set the global OMEMO trust mode to ToFu ( Trust on first use ), first OMEMO keys trusted automatic.
3. trustmode blind - Set the global OMEMO trust mode to Blind, ALL OMEMO keys trusted automatic.

### CODE REVIEW

**Comment by StefanKropp** [comment]
There is a loop during blind trust because of 
https://github.com/profanity-im/profanity/blob/master/src/omemo/omemo.c#L1169

File: src/omemo/omemo.c [hunk 1]
```diff
@@ -414,6 +414,36 @@ omemo_start_session(const char* const barejid)
...
+                    cons_show("OMEMO: No trusted devices found for %s", barejid);
+                    GList* device_id;
+                    for (device_id = device_list; device_id != NULL; device_id = device_id->next) {
+                        GHashTable* known_identities = g_hash_table_lookup(omemo_ctx.known_devices, barejid);
+                        if (known_identities) {
+                            GList *fp = g_hash_table_get_keys (known_identities);
+                            for ( fp = g_hash_table_get_keys (known_identities); fp != NULL; fp = fp->next ) {
+                                if ( device_id->data == g_hash_table_lookup(known_identities, fp->data) ) {
+                                    cons_show("OMEMO: Adding ToFu trust for %s device %d - Fingerprint %s", barejid, device_id->data, omemo_format_fingerprint( fp->data ));
+                                    omemo_trust(barejid, omemo_format_fingerprint( fp->data ));
```
Comment (paulfariello) [inline src/omemo/omemo.c:439]: We could add another `omemo_trust` function to avoid this useless fingerprint formating.

**Review by paulfariello — CHANGES_REQUESTED** [review]

**Review by jubalh — CHANGES_REQUESTED** [review]
See pauls comments.
Plus maybe some commit descriptions :)

**Review by mwuttke97 — CHANGES_REQUESTED** [review]
> first OMEMO keys trusted automatic

There's a typo, and it's also too vague. I think the new feature should also be documented in the user guide.

**Comment by jubalh** [comment]
Please make sure to answer on the above review from paul if necessary or click on resolved or request a new round of reviews once you are ready. I'm not sure if your new commits are just for you or you already look for a next review round :)

**Comment by kaffeekanne** [comment]
I would rename `tofo` to `firstusage`/`firstsight` or something like that. Although profanity is a client targeted for intermediate (at least) users we have mostly descriptive language in the interface. We use the term `rooms` instead of `MUC` for example.

File: src/config/preferences.c [hunk 2]
```diff
@@ -2320,6 +2323,8 @@ _get_default_string(preference_t pref)
         return "redact";
     case PREF_OMEMO_POLICY:
         return "automatic";
+    case PREF_OMEMO_TRUST_MODE:
+        return "manuel";
```
[the remark this answers is not in the document]
Comment (StefanKropp) [reply src/config/preferences.c]: ACK, solved with commit: c503a70d78af82bff16dd919d19467c71ed093f6

File: src/omemo/omemo.c [hunk 3]
```diff
@@ -414,6 +414,36 @@ omemo_start_session(const char* const barejid)
         for (device_id = device_list; device_id != NULL; device_id = device_id->next) {
             omemo_bundle_request(barejid, GPOINTER_TO_INT(device_id->data), omemo_start_device_session_handle_bundle, free, strdup(barejid));
         }
+
+        // OMEMO trustmode ToFu
+        if (g_strcmp0(prefs_get_string(PREF_OMEMO_TRUST_MODE), "tofu") == 0) {
+            log_info("[OMEMO] Checking ToFu state for %s", barejid);
+            GHashTable* trusted = trusted = g_hash_table_lookup(omemo_ctx.identity_key_store.trusted, barejid);
```
[the remark this answers is not in the document]
Comment (StefanKropp) [reply src/omemo/omemo.c]: ACK, thanks. Solved in 8cfc9903fc75f00c376d943d3809f206a1cf92f3

File: src/omemo/omemo.c [hunk 4]
```diff
@@ -414,6 +414,36 @@ omemo_start_session(const char* const barejid)
...
+            if (trusted) {
+                if ( g_hash_table_size(trusted) > 0 ) {
+                    log_info("[OMEMO] Found trusted device for %s - skip ToFu", barejid);
+                    return;
+                }
+            } else {
+                GList* device_list = g_hash_table_lookup(omemo_ctx.device_list, barejid);
```
[the remark this answers is not in the document]
Comment (StefanKropp) [reply src/omemo/omemo.c]: Done, thanks. Solved in 38d0b2dc5eec68e207fc90a071f8afbd712c0861

File: src/omemo/omemo.c (hunk 1 above)
Comment (StefanKropp) [reply src/omemo/omemo.c]: Not done,... because of time. We can do some more improvements later.

**Comment by StefanKropp** [comment]
> I would rename `tofo` to `firstusage`/`firstsight` or something like that. Although profanity is a client targeted for intermediate (at least) users we have mostly descriptive language in the interface. We use the term `room` instead of `MUC` for example.

OK,.. I will change to `firstusage`

**Comment by StefanKropp** [comment]
> > I would rename `tofo` to `firstusage`/`firstsight` or something like that. Although profanity is a client targeted for intermediate (at least) users we have mostly descriptive language in the interface. We use the term `room` instead of `MUC` for example.
> 
> OK,.. I will change to `firstusage`

4f50f8d85eb4e1977813ef1fb3dcf7c65af127cf

File: src/omemo/omemo.c [hunk 5]
```diff
@@ -414,6 +414,36 @@ omemo_start_session(const char* const barejid)
...
+                if (device_list) {
+                    cons_show("OMEMO: No trusted devices found for %s", barejid);
+                    GList* device_id;
+                    for (device_id = device_list; device_id != NULL; device_id = device_id->next) {
+                        GHashTable* known_identities = g_hash_table_lookup(omemo_ctx.known_devices, barejid);
+                        if (known_identities) {
+                            GList *fp = g_hash_table_get_keys (known_identities);
```
[the remark this answers is not in the document]
Comment (StefanKropp) [reply src/omemo/omemo.c]: Fixed in 8cfc9903fc75f00c376d943d3809f206a1cf92f3

File: src/omemo/omemo.c [hunk 6]
```diff
@@ -619,6 +647,13 @@ omemo_start_device_session(const char* const jid, uint32_t device_id,
 
     gboolean trusted = is_trusted_identity(&address, (uint8_t*)identity_key_raw, identity_key_len, &omemo_ctx.identity_key_store);
 
+    if ((g_strcmp0(prefs_get_string(PREF_OMEMO_TRUST_MODE), "blind") == 0) && !trusted) {
+        char* fp = _omemo_fingerprint(identity_key, TRUE);
```
Comment (jubalh) [inline src/omemo/omemo.c:651]: AFAIk fp needs to be freed.

File: src/config/preferences.c [hunk 7]
```diff
@@ -2320,6 +2323,8 @@ _get_default_string(preference_t pref)
         return "redact";
     case PREF_OMEMO_POLICY:
         return "automatic";
+    case PREF_OMEMO_TRUST_MODE:
```
Comment (jubalh) [inline src/config/preferences.c:2326]: that's like it is now, right?

File: src/command/cmd_funcs.c [hunk 8]
```diff
@@ -8489,6 +8489,37 @@ cmd_omemo_start(ProfWin* window, const char* const command, gchar** args)
...
+#ifdef HAVE_OMEMO
+
+    if (!args[1]) {
+        cons_show("Current trust mode is %s", prefs_get_string(PREF_OMEMO_TRUST_MODE));
+        return TRUE;
+    }
+
+    if (g_strcmp0(args[1], "manual") == 0) {
+        cons_show("Current trust mode is %s - setting to %s", prefs_get_string(PREF_OMEMO_TRUST_MODE), args[1]);
+        cons_show("You need to trust all OMEMO fingerprint manually");
```
Comment (jubalh) [inline src/command/cmd_funcs.c:8504]: s/fingerprint/fingerprints

File: src/command/cmd_defs.c [hunk 9]
```diff
@@ -2313,6 +2315,9 @@ static struct cmd_t command_defs[] = {
             { "log redact",              "Log OMEMO encrypted messages, but replace the contents with [redacted]. This is the default." },
             { "fingerprint [<contact>]", "Show contact fingerprints, or current recipient if omitted." },
             { "char <char>",             "Set the character to be displayed next to OMEMO encrypted messages." },
+            { "trustmode manual",        "Set the global OMEMO trust mode to manual, OMEMO keys has to be trusted manually." },
+            { "trustmode firstusage",    "Set the global OMEMO trust mode to ToFu, first OMEMO keys trusted automatic." },
+            { "trustmode blind",         "Set the global OMEMO trust mode to Blind, ALL OMEMO keys trusted automatic." },
```
Comment (jubalh) [inline src/command/cmd_defs.c:2320]: s/Blind/blind
s/automatic/automatically

File: src/command/cmd_defs.c [hunk 10]
```diff
@@ -2313,6 +2315,9 @@ static struct cmd_t command_defs[] = {
             { "log redact",              "Log OMEMO encrypted messages, but replace the contents with [redacted]. This is the default." },
             { "fingerprint [<contact>]", "Show contact fingerprints, or current recipient if omitted." },
             { "char <char>",             "Set the character to be displayed next to OMEMO encrypted messages." },
+            { "trustmode manual",        "Set the global OMEMO trust mode to manual, OMEMO keys has to be trusted manually." },
+            { "trustmode firstusage",    "Set the global OMEMO trust mode to ToFu, first OMEMO keys trusted automatic." },
```
Comment (jubalh) [inline src/command/cmd_defs.c:2319]: s/automatic/automatically

**Review by jubalh — CHANGES_REQUESTED** [review]
See comments :)

**Comment by jubalh** [comment]
Would be cool if someone could test this and report back here.

File: src/config/preferences.c (hunk 7 above)
Comment (StefanKropp) [reply src/config/preferences.c:2326]: > that's like it is now, right?

Yes

File: src/command/cmd_defs.c (hunk 9 above)
Comment (StefanKropp) [reply src/command/cmd_defs.c]: fixed

File: src/command/cmd_defs.c (hunk 10 above)
Comment (StefanKropp) [reply src/command/cmd_defs.c]: fixed

File: src/omemo/omemo.c (hunk 6 above)
Comment (StefanKropp) [reply src/omemo/omemo.c:651]: done and tested

**Review by jubalh — APPROVED** [review]

**Pull request merged**